src/libraries/Common/src/System/HexConverter.cs

```diff
@@ -261,7 +261,13 @@ public static bool TryDecodeFromUtf16_Vector128(ReadOnlySpan<char> chars, Span<b
                 // single UTF8 ASCII vector - the implementation can be shared with UTF8 paths.
                 Vector128<ushort> vec1 = Vector128.LoadUnsafe(ref srcRef, offset);
                 Vector128<ushort> vec2 = Vector128.LoadUnsafe(ref srcRef, offset + (nuint)Vector128<ushort>.Count);
-                Vector128<byte> vec = Vector128.Narrow(vec1, vec2);
+
+                // Pack the two UTF-16 vectors into a single UTF8 ASCII vector.
+                // PackUnsignedSaturate is cheaper than Vector128.Narrow on X86, but still correct in this case as
+                // any values that aren't valid ASCII will also not be valid Hex after narrowing, and get rejected below.
+                Vector128<byte> vec = Sse2.IsSupported
```

Comment:
We're trying to move away from xplat intrinsics where possible, not more towards them.

This is one of those cases where I think if there is enough perf advantage then we should look at having an additional helper API instead. -- We have `Narrow` and approved but haven't yet implemented `NarrowWithSaturation`, this is a case where the latter would be sufficient in most cases but where it's not faster than regular `Narrow` on some newer hardware where dedicated instructions exist.

Reply:
This could be using the helper we have in Ascii https://github.com/dotnet/runtime/blob/b26d11c1ee3d99ebdbb837408c60262f5113cdab/src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Utility.cs#L1655
Are you suggesting we expose that as another helper - e.g. `NarrowUnsafe`, akin to `ShuffleUnsafe` where out-of-range values are UB?

Reply:
We couldn't use a `NarrowUnsafe` here since `UB` would include such out of range values becoming `zero`, which would not be valid for this scenario.

This is more rather that we really want to minimize the usage of platform specific intrinsics where possible and where they do get used, we really want to abstract them into some helper (even if its specific to the one method/spot and not more broadly reusable).

We can't maintain the code as easily if we start adding in platform specific optimizations everywhere, especially as the set of architectures/ISAs we need to care about increases (Arm64 vs Xarch vs Wasm vs LA64 vs RV64 vs ...). It also gets complicated because there's places where opportunistic light up like this is can be worse on newer hardware. For example, SSE2 `packus*, which saturates,` is often slightly slower (although not meaningfully so) than AVX512 `vpmov*`, which truncates.

Reply:
Yeah, just move it to a helper method here given its a 33% perf increase in some cases.

> Going to zero should be fine in this case, it's what's happening with PackUnsignedSaturate.

Sorry, my point was that `Unsafe` means that any behavior could happen. It could even saturate to a valid ascii character like `'a'`. If we actually don't care what the value is, because something else masks it or similar, then that's fine, but wasn't how I interpreted the comments around the code.

Reply:
I see this part in the HexConverter is already under an `#if SYSTEM_PRIVATE_CORELIB` condition, so we can use the existing internal helper from Ascii utilities here.

> It also gets complicated because there's places where opportunistic light up like this is can be worse on newer hardware

`Ascii.Utility.cs` has a bunch of these we might want to look at at some point